File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.ts
function maxProfit(prices: number[]): number {
    let minPrice: number = Infinity
    let maxProfit: number = 0

    for (const price of prices) {
        if (price < minPrice) {
            minPrice = price
        } else {
            const profit: number = price - minPrice
            if (profit > maxProfit) {
                maxProfit = profit
            }
        }
    }

    return maxProfit
};

File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        max_profit = 0
        min_buying_price = prices[0]

        for price in prices:
            curr_profit = price - min_buying_price
            min_buying_price = min(min_buying_price, price)
            max_profit = max(max_profit, curr_profit)

        return max_profit